### PR TITLE
Add Hometown to Submission Fields

### PR DESCRIPTION
**Pull request by johnbyrne727rit**
This PR adds the hometown field to show submissions. If a user does not have a hometown set already, the field is enabled. Otherwise, it displays their hometown and is disabled.

See [Trello Card](https://trello.com/c/GNjxupje/56-add-hometown-to-submission-field).

### CODE REVIEW

File: backend/db/migrations/20191103153600-add-hometown-to-users.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+export function up (queryInterface, Sequelize) {
```
Comment (johnbyrne727rit) [inline backend/db/migrations/20191103153600-add-hometown-to-users.js:1]: This file is the migration that adds hometown to the user table.

File: backend/models/user.js [hunk 2]
```diff
@@ -21,6 +21,10 @@ const User = sequelize.define('user', {
     allowNull: false,
     notEmpty: true
   },
+  hometown: {
+    type: DataTypes.STRING,
```
Comment (johnbyrne727rit) [inline backend/models/user.js:25]: This is adding hometown to the backend model.

File: backend/resolvers/mutations/entry.js [hunk 3]
```diff
@@ -28,17 +29,38 @@ const createEntry = (entry, entryType, entryId, t) => {
...
-        ...newEntry,
-        entryType: entryType,
-        entryId: entryId,
-        groupId: group ? group.id : null
-      }, { transaction: t })
+      return userFindPromise.then(user => {
+
+        let userUpdatePromise = Promise.resolve(null)
+        if (user.hometown != entry.hometown){
+          userUpdatePromise = User.update(
```
Comment (johnbyrne727rit) [inline backend/resolvers/mutations/entry.js:46]: Updates the hometown on the user if necessary.

File: backend/schema.js [hunk 4]
```diff
@@ -15,6 +15,7 @@ type User {
     firstName: String!
     lastName: String!
     displayName: String
+    hometown: String
```
Comment (johnbyrne727rit) [inline backend/schema.js:18]: So backend knows to send hometown to the frontend with the user.

File: backend/schema.js [hunk 5]
```diff
@@ -115,6 +116,7 @@ input EntryInput {
     title: String!
     comment: String
     forSale: Boolean
+    hometown: String
```
Comment (johnbyrne727rit) [inline backend/schema.js:119]: So the backend knows to expect the hometown when an entry is created.

File: frontend/src/Student/components/HometownInput.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import React from "react";
+import { Label, FormGroup } from "reactstrap";
+import { Field } from "formik";
+
+export default function HometownInput({
```
Comment (johnbyrne727rit) [inline frontend/src/Student/components/HometownInput.js:5]: A React component for the Hometown input box.

File: frontend/src/Student/containers/OtherMediaSubmissionForm.js [hunk 7]
```diff
@@ -19,6 +19,7 @@ const mapDispatchToProps = dispatch => ({
   done: () => dispatch(push('/')),
   handlePDFUpload: file => dispatch(uploadPDF(file)),
   handleImageUpload: file => dispatch(uploadImage(file)),
+  handleHometown: hometown => dispatch(setUserHometown(hometown)),
```
Comment (johnbyrne727rit) [inline frontend/src/Student/containers/OtherMediaSubmissionForm.js:22]: Passing the action to the component.

File: frontend/src/Student/components/VideoSubmissionForm.js [hunk 8]
```diff
@@ -161,6 +168,9 @@ class VideoSubmissionForm extends Component {
 
             // Create an entry, show the success modal, and then go to the dashboard
             create(input)
+              .then(()=>{
+                handleHometown(values.hometown)
```
Comment (johnbyrne727rit) [inline frontend/src/Student/components/VideoSubmissionForm.js:172]: After successful creation, update the front end hometown

File: frontend/src/shared/actions.js [hunk 9]
```diff
@@ -79,3 +80,10 @@ export const dismissError = index => (dispatch, getState, client) => {
     payload: index
   })
 }
+
+export const setUserHometown = hometown => (dispatch, getState, client) => {
+  dispatch({
```
Comment (johnbyrne727rit) [inline frontend/src/shared/actions.js:85]: New action for hometown setting.

File: frontend/src/shared/reducers.js [hunk 10]
```diff
@@ -50,6 +51,17 @@ const auth = (state = {}, action) => {
         ...state,
         downloadToken: action.payload
       }
+    case actions.SET_USER_HOMETOWN:
+        if(state.user.hometown){
+          return state;
+        }
```
Comment (johnbyrne727rit) [inline frontend/src/shared/reducers.js:57]: Only alter the state if there is no hometown already.

**Comment by JohnPMurray** [comment]
Is there any way at all to change a users hometown once its created? I know we didn't really decide on how to go about that, but we did agree that it should be changeable even if its just by the admin. 
Maybe this is something we can put off until later, but we should add a card for it in that case.

**Comment by johnbyrne727rit** [comment]
@JohnPMurray As of right now there isn't. On the front end side, the input box is disabled once the hometown is created. If someone messes with the inspector to alter the value anyway the form will let the request go through, but the code on backend and frontend that alters the hometown won't let it through.

Here's the block on the FE
https://github.com/johnbyrne727rit/gallery-gateway/blob/662424b7eed824da1caffd82f710f5bb40238525/frontend/src/shared/reducers.js#L55

And here it is on the BE
https://github.com/johnbyrne727rit/gallery-gateway/blob/662424b7eed824da1caffd82f710f5bb40238525/backend/resolvers/mutations/entry.js#L45

**Review by JohnPMurray — APPROVED** [review]

**Review by moiseslorap — APPROVED** [review]

**Pull request merged**